### PR TITLE
feat(automatic_goal): add automatic goal rviz plugin

### PR DESCRIPTION
**Pull request by dmoszynski**
## Description
The `tier4_automatic_goal_rviz_plugin` plugin has been developed.
It is designed to automatically achieve subsequent goals.
### Plugin appearance
counting_panel
### Plugin features
- Defining a `GoalsList`  by adding goals using `RvizTool` (`Pose` on the map).
- Automatic execution of the created `GoalsList`  from the selected goal - it can be stopped and restarted.
- Looping the current `GoalsList` .
- Saving achieved goals to a file.
- Plan the route to one (single) selected goal and starting that route - it can be stopped and restarted.
- Remove any goal from the list or clear the current route.
- Save the current `GoalsList`  to a file and load the list from the file.
- The application enables/disables access to options depending on the current state
- The saved `GoalsList` can be executed without using a plugin - using a node `automatic_goal_sender`
    - `ros2 launch tier4_automatic_goal_rviz_plugin automatic_goal_sender.launch.xml goals_list_file_path:="/home/ros2/goals_list.yaml" goals_achieved_dir_path:="/home/ros2/"`

## Tests performed

All plugin features have been tested manually.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [ ] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by shmpwk** [comment]
@dmoszynski 
Thank you for your contribution, could you check the CI?

**Comment by taikitanaka3** [comment]
@dmoszynski 
Thank you for nice PR.
Can you add @shmpwk as a code maintainer and you can add author by your name.  
```
  <maintainer email="shumpei.wakabayashi@tier4.jp">Shumpei Wakabayashi</maintainer>

  <author email="hogehoge"> your name</author>
```

File: common/tier4_automatic_goal_rviz_plugin/package.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>tier4_automatic_goal_rviz_plugin</name>
+  <version>0.0.0</version>
+  <description>The autoware automatic goal rviz plugin package</description>
+  <maintainer email="shumpei.wakabayashi@tier4.jp">Shumpei Wakabayashi</maintainer>
+  <author email="dawid.moszynski@robotec.ai">Dawid Moszyński</author>
```
Comment (taikitanaka3) [inline common/tier4_automatic_goal_rviz_plugin/package.xml:8]: Thank you for adding and please add you to maintainer too.
```suggestion
   <maintainer email="dawid.moszynski@robotec.ai">Dawid Moszyński</maintainer>
  <author email="dawid.moszynski@robotec.ai">Dawid Moszyński</author>
```

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_append_tool.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 Tier IV, Inc.
```
Comment (satoshi-ota) [inline common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_append_tool.cpp:1]: [nits]
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_append_tool.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 Tier IV, Inc.
```
Comment (satoshi-ota) [inline common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_append_tool.hpp:1]:
```suggestion
// Copyright 2023 TIER IV, Inc.
```

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.hpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
+//
+//  Copyright 2020 Tier IV, Inc. All rights reserved.
```
Comment (satoshi-ota) [inline common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.hpp:2]:
```suggestion
//  Copyright 2023 TIER IV, Inc. All rights reserved.
```

**Comment by shmpwk** [comment]
@dmoszynski 
Could you check the failed test?
https://github.com/autowarefoundation/autoware.universe/actions/runs/4382216946/jobs/7674202659

**Comment by dmoszynski** [comment]
failed as result because:
```
4: File 'package.xml' is invalid:
4: /__w/autoware.universe/autoware.universe/common/tier4_automatic_goal_rviz_plugin/package.xml:9: element author: Schemas validity error : Element 'author': This element is not expected. Expected is one of ( maintainer, license ).
4: /__w/autoware.universe/autoware.universe/common/tier4_automatic_goal_rviz_plugin/package.xml fails to validate
4: 
```
I'm not sure what I can do. Should I remove the `<author>`? @taikitanaka3 suggested to add this. (https://github.com/autowarefoundation/autoware.universe/pull/3031#issuecomment-1462087308)

**Comment by shmpwk** [comment]
@dmoszynski 
How about adding LICENSE before author tag?

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_sender.hpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+  // Cli
+  rclcpp::Client<ChangeOperationMode>::SharedPtr cli_change_to_autonomous_;
+  rclcpp::Client<ChangeOperationMode>::SharedPtr cli_change_to_stop_;
+  rclcpp::Client<ClearRoute>::SharedPtr cli_clear_route_;
+  rclcpp::Client<SetRoutePoints>::SharedPtr cli_set_route_;
+
+  // Containers
+  unsigned current_goal{0};
+  State state{State::INITIALIZING};
```
Comment (taikitanaka3) [inline common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_sender.hpp:157]: this current_goal and state should  be current_goal_ and state_  and please fix other part as well.

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.cpp [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+  QString file_name = QFileDialog::getOpenFileName(
+    this, tr("Open File with GoalsList"), "/home", tr("Goal lists (*.yaml)"));
+  if (file_name.count() > 0) loadGoalsList(file_name.toStdString());
+}
+
+void AutowareAutomaticGoalPanel::onClickSaveListToFile()
+{
+  if (goals_list_.size() > 0) {
+    QString file_name = QFileDialog::getSaveFileName(
+      this, tr("Save File with  GoalsList"), "/home/goals_list.yaml", tr("Goal lists (*.yaml)"));
```
Comment (taikitanaka3) [inline common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.cpp:238]: `/home/goals_list.yaml` I think this kind of file path should be `log/` or `/tmp/hoge`

**Comment by taikitanaka3** [comment]
@dmoszynski 
I tried to test this PR with readme description but I couldn't use as expected because I dont have file of goal_list.yaml yet.
image

`ros2 launch tier4_automatic_goal_rviz_plugin automatic_goal_sender.launch.xml goals_list_file_path:="/home/ros2/goals_list.yaml" goals_achieved_dir_path:="/home/ros2/"`

and this gives me error `goals_list.yaml` is not found.

File: common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.cpp (hunk 6 above)
Comment (taikitanaka3) [reply common/tier4_automatic_goal_rviz_plugin/src/automatic_goal_panel.cpp:238]: similar discussion
https://github.com/autowarefoundation/autoware/pull/3031

**Comment by dmoszynski** [comment]
@taikitanaka3
Node `automatic goal_sender` is only useful if you want to run a previously created and saved `GoalsList`.
Please do this:
1. Add `AutomaticGoalTool` - to allow you to append your goals to `GoalsList`.
2. Append multiple goals one by one - in such a way that routes can be planned.
3. You can try `GoalsList` by clicking `Send goals automatically`.
4. You can save  `GoalsList` by clicking `Save to file`.
5. After saving, you can run the `GoalsList` without using a plugin - using the command you mentioned.
`goals_achieved_dir_path` - is the path to the directory where the file `goals_achieved.log` will be created and tthe achieved goals will be written to it

**Comment by taikitanaka3** [comment]
@dmoszynski 
Thank you I found a AutomaticGoalTool on Rviz tool but this time I met this error when adding this AutomaticGoalPanel
It looks like some error occurs at making instance of AutomaticGoalPanel.

Can you test this panel again?

```
#9  0x00007fbb59aa538a in std::__throw_logic_error(char const*) () at /lib/x86_64-linux-gnu/libstdc++.so.6
#10 0x00007fbae4a2cbd2 in automatic_goal::AutowareAutomaticGoalSender::AutowareAutomaticGoalSender() () at /home/t4tanaka/workspace/autoware/install/tier4_automatic_goal_rviz_plugin/lib/libtier4_automatic_goal_rviz_plugin.so
#11 0x00007fbae4a7ee36 in rviz_plugins::AutowareAutomaticGoalPanel::AutowareAutomaticGoalPanel(QWidget*) () at /home/t4tanaka/workspace/autoware/install/tier4_automatic_goal_rviz_plugin/lib/libtier4_automatic_goal_rviz_plugin.so
#12 0x00007fbae4a8bd92 in class_loader::impl::MetaObject<rviz_plugins::AutowareAutomaticGoalPanel, rviz_common::Panel>::create() const ()
    at /home/t4tanaka/workspace/autoware/install/tier4_automatic_goal_rviz_plugin/lib/libtier4_automatic_goal_rviz_plugin.so

```

**Comment by dmoszynski** [comment]
@taikitanaka3 
I'm really sorry! My `ref` change today caused this - already fixed and tested with and without the plugin as well.
I also expanded the `readme` a bit, maybe it will be helpful.

**Review by taikitanaka3 — APPROVED** [review]
@dmoszynski 
next time when you make PR  it would be better not to make rebase commit and please make a merge commit, because it's hard to trace change history with rebase commit every time.


LGTM for this PR
image (2)

File: common/tier4_automatic_goal_rviz_plugin/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+2. Select `tier4_automatic_goal_rviz_plugin/AutowareAutomaticGoalPanel` and press OK.
+
+3. Select Add a new tool.
+
+   ![select_tool](./images/select_tool.png)
+
+4. Select `tier4_automatic_goal_rviz_plugin/AutowareAutomaticGoalTool` and press OK.
+
+5. Add goals visualiization as markers to `Displays`.
```
Comment (satoshi-ota) [inline common/tier4_automatic_goal_rviz_plugin/README.md:57]: Could you fix this typo? :pray:

Comment (satoshi-ota) [reply common/tier4_automatic_goal_rviz_plugin/README.md:57]:
```suggestion
5. Add goals visualization as markers to `Displays`.
```

Comment (dmoszynski) [reply common/tier4_automatic_goal_rviz_plugin/README.md]: Oh sorry for the typo, thank you very much!

**Comment by dmoszynski** [comment]
@taikitanaka3 
Thank you so much! Please forgive me for so many rebases and changes. I tried to keep everything clear.
I need to develop a habit of signing commits.

**Pull request merged**